### DISCUSSION TITLE
Can TypeVar defaults be used to inform Pyright in cases where an argument isn't provided?

### DISCUSSION THREAD

**Discussion by max-muoto** [microsoft/pyright#8084] [Q&A]
One thing I was curious about with the new 3.13 type var defaults, is if they could be used to inform the type-checker in the case of an argument that might not be provided:

```python
T = TypeVar("T", bound=str, default=str)

def foo(arg_1: int, *, arg_2: T = ...) -> T: ...
```

[Playground](https://pyright-play.net/?pyrightVersion=1.1.353&pythonVersion=3.13&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgGoCGIAUDQVALyEkUgAUARAewDRQBGYAK4oAJvQDOMELxHFg5QQBsYEqQEoas4FGBgwrSmgD6ARgBcmFDF4AqXoaMAmC3UYA6D2qgBaAHyELDzcgA)

Currently this raises a warning, but I wasn't sure based on the language of PEP 696 if this has to be the case. 

> In generic functions, type checkers may use a type parameter’s default when the type parameter cannot be solved to anything. We leave the semantics of this usage unspecified, as ensuring the default is returned in every code path where the type parameter can go unsolved may be too hard to implement. Type checkers are free to either disallow this case or experiment with implementing support.

I feel this gives Pyright flexibility to support this case if it wants. Personally, I think there are a few situations where this helpful, or at least running into one right now.

**Comment by max-muoto** [comment]
For context, this is the behavior in MyPy. [MyPy playground](https://mypy-play.net/?mypy=latest&python=3.12&flags=verbose%2Cignore-missing-imports%2Cshow-error-context%2Cimplicit-optional%2Callow-untyped-globals%2Cstrict%2Ccheck-untyped-defs%2Cdisallow-any-decorated%2Cdisallow-any-expr%2Cdisallow-any-generics%2Cdisallow-any-unimported%2Cdisallow-incomplete-defs%2Cdisallow-subclassing-any%2Cdisallow-untyped-calls%2Cdisallow-untyped-decorators%2Cdisallow-untyped-defs%2Cno-implicit-reexport%2Clocal-partial-types%2Cno-strict-optional%2Cno-warn-no-return%2Cstrict-equality%2Cwarn-incomplete-stub%2Cwarn-redundant-casts%2Cwarn-return-any%2Cwarn-unreachable%2Cwarn-unused-configs).

**Reply by erictraut** [reply]
When you say "an argument that isn't provided", do you literally mean an argument that is specified as `...` (an ellipsis) as is sometimes done in stub files? Or are you referring to the case where the real default argument expression is provided, as in:

```python
def foo(arg_1: int, *, arg_2: T = "default") -> T: ...
```

If you provide the actual default argument expression, pyright's warning will go away.

The type of `...` is not necessarily the same as the type of the default value. Assuming that they are the same type is a bad assumption that will lead to incorrect type evaluations. Here's an example of where this assumption fails.

```python
from typing import TypeVar

T = TypeVar("T", default=str)

def foo(arg_1: int, *, arg_2: T = ...) -> T: ... # Let's say that the default value for arg_2 is `None` at runtime

reveal_type(foo(1)) # This should be `None`, but it is `str` because of the bad assumption
```

In summary, I think pyright is correct to emit a warning in the case where a `...` is used for a parameter whose annotated type is a TypeVar. This combination is unsafe even when the TypeVar has a default type.

**Reply by max-muoto** [reply]
Really just in the case of overloads and stubs, or at least that's where I'm encountering this issue.

**Reply by max-muoto** [reply]
Makes sense for this behavior to be the case for an actual function of course, but in those two cases it seems more like a false positive to me.

**Reply by max-muoto** [reply]
As for why I wouldn't just set a default in the stubs, I'm making some changes to stubs in the typeshed, and they have some linting rules to only allow "simple" defaults.

**Reply by max-muoto** [reply]
The PR if you're curious: https://github.com/python/typeshed/pull/12104

**Reply by max-muoto** [reply]
@erictraut Just wanted to follow up here. Is there possibility of cutting an exception in the case of stubs?

**Reply by erictraut** [reply]
No, even if this behavior were sound from a type checking standpoint (which it isn't), there would be no justification for special-casing stubs in this case.

**Reply by max-muoto** [reply]
> No, even if this behavior were sound from a type checking standpoint (which it isn't), there would be no justification for special-casing stubs in this case.

Sounds good!
